File: qt/gen_qstringconverter.go
package qt

import (
	"unsafe"
)

type QStringEncoder struct {
	h          uintptr
	isSubclass bool
}

// NewQStringEncoder constructs a new QStringEncoder object.
func NewQStringEncoder() *QStringEncoder {

	ret := newQStringEncoder(QStringEncoder_new())
	ret.isSubclass = true
	return ret
}

// NewQStringEncoder2 constructs a new QStringEncoder object.
func NewQStringEncoder2(encoding Encoding) *QStringEncoder {

	ret := newQStringEncoder(QStringEncoder_new2(encoding))
	ret.isSubclass = true
	return ret
}

// NewQStringEncoder3 constructs a new QStringEncoder object.
func NewQStringEncoder3(name QAnyStringView) *QStringEncoder {

	ret := newQStringEncoder(QStringEncoder_new3(name.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQStringEncoder4 constructs a new QStringEncoder object.
func NewQStringEncoder4(encoding Encoding, flags Flags) *QStringEncoder {

	ret := newQStringEncoder(QStringEncoder_new4(encoding, flags))
	ret.isSubclass = true
	return ret
}

// NewQStringEncoder5 constructs a new QStringEncoder object.
func NewQStringEncoder5(name QAnyStringView, flags Flags) *QStringEncoder {

	ret := newQStringEncoder(QStringEncoder_new5(name.cPointer(), flags))
	ret.isSubclass = true
	return ret
}

func (this *QStringEncoder) RequiredSpace(inputLength int64) int64 {
	return (int64)(QStringEncoder_RequiredSpace(this.h, (ptrdiff_t)(inputLength)))
}

type QStringDecoder struct {
	h          uintptr
	isSubclass bool
}

// NewQStringDecoder constructs a new QStringDecoder object.
func NewQStringDecoder(encoding Encoding) *QStringDecoder {

	ret := newQStringDecoder(QStringDecoder_new(encoding))
	ret.isSubclass = true
	return ret
}

// NewQStringDecoder2 constructs a new QStringDecoder object.
func NewQStringDecoder2() *QStringDecoder {

	ret := newQStringDecoder(QStringDecoder_new2())
	ret.isSubclass = true
	return ret
}

// NewQStringDecoder3 constructs a new QStringDecoder object.
func NewQStringDecoder3(name QAnyStringView) *QStringDecoder {

	ret := newQStringDecoder(QStringDecoder_new3(name.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQStringDecoder4 constructs a new QStringDecoder object.
func NewQStringDecoder4(encoding Encoding, flags Flags) *QStringDecoder {

	ret := newQStringDecoder(QStringDecoder_new4(encoding, flags))
	ret.isSubclass = true
	return ret
}

// NewQStringDecoder5 constructs a new QStringDecoder object.
func NewQStringDecoder5(name QAnyStringView, f Flags) *QStringDecoder {

	ret := newQStringDecoder(QStringDecoder_new5(name.cPointer(), f))
	ret.isSubclass = true
	return ret
}

func (this *QStringDecoder) RequiredSpace(inputLength int64) int64 {
	return (int64)(QStringDecoder_RequiredSpace(this.h, (ptrdiff_t)(inputLength)))
}

func (this *QStringDecoder) AppendToBuffer(out *QChar, ba QByteArrayView) *QChar {
	return newQChar(QStringDecoder_AppendToBuffer(this.h, out.cPointer(), ba.cPointer()))
}

func QStringDecoder_DecoderForHtml(data QByteArrayView) *QStringDecoder {
	_goptr := newQStringDecoder(QStringDecoder_DecoderForHtml(data.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

File: qt/gen_qstringconverter.cpp
// +build ignore

#include <QAnyStringView>
#include <QByteArrayView>
#include <QChar>
#include <QStringConverter>
#include <QStringConverterBase>
#include <QStringDecoder>
#include <QStringEncoder>
#include <qstringconverter.h>
#include "gen_qstringconverter.h"

QStringEncoder* QStringEncoder_new() {
	return new QStringEncoder();
}

QStringEncoder* QStringEncoder_new2(Encoding encoding) {
	return new QStringEncoder(encoding);
}

QStringEncoder* QStringEncoder_new3(QAnyStringView* name) {
	return new QStringEncoder(*name);
}

QStringEncoder* QStringEncoder_new4(Encoding encoding, Flags flags) {
	return new QStringEncoder(encoding, flags);
}

QStringEncoder* QStringEncoder_new5(QAnyStringView* name, Flags flags) {
	return new QStringEncoder(*name, flags);
}

void QStringEncoder_virtbase(QStringEncoder* src, QStringConverter** outptr_QStringConverter) {
	*outptr_QStringConverter = static_cast<QStringConverter*>(src);
}

ptrdiff_t QStringEncoder_RequiredSpace(const QStringEncoder* self, ptrdiff_t inputLength) {
	qsizetype _ret = self->requiredSpace((qsizetype)(inputLength));
	return static_cast<ptrdiff_t>(_ret);
}

void QStringEncoder_Delete(QStringEncoder* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QStringEncoder*>( self );
	} else {
		delete self;
	}
}

QStringDecoder* QStringDecoder_new(Encoding encoding) {
	return new QStringDecoder(encoding);
}

QStringDecoder* QStringDecoder_new2() {
	return new QStringDecoder();
}

QStringDecoder* QStringDecoder_new3(QAnyStringView* name) {
	return new QStringDecoder(*name);
}

QStringDecoder* QStringDecoder_new4(Encoding encoding, Flags flags) {
	return new QStringDecoder(encoding, flags);
}

QStringDecoder* QStringDecoder_new5(QAnyStringView* name, Flags f) {
	return new QStringDecoder(*name, f);
}

void QStringDecoder_virtbase(QStringDecoder* src, QStringConverter** outptr_QStringConverter) {
	*outptr_QStringConverter = static_cast<QStringConverter*>(src);
}

ptrdiff_t QStringDecoder_RequiredSpace(const QStringDecoder* self, ptrdiff_t inputLength) {
	qsizetype _ret = self->requiredSpace((qsizetype)(inputLength));
	return static_cast<ptrdiff_t>(_ret);
}

QChar* QStringDecoder_AppendToBuffer(QStringDecoder* self, QChar* out, QByteArrayView* ba) {
	return self->appendToBuffer(out, *ba);
}

QStringDecoder* QStringDecoder_DecoderForHtml(QByteArrayView* data) {
	return new QStringDecoder(QStringDecoder::decoderForHtml(*data));
}

void QStringDecoder_Delete(QStringDecoder* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QStringDecoder*>( self );
	} else {
		delete self;
	}
}

